singer-commons/src/main/java/com/pinterest/singer/loggingaudit/client/LoggingAuditClient.java

```diff
@@ -0,0 +1,258 @@
+  }
+
+
+  public LoggingAuditClientConfig getConfig() {
+    return config;
+  }
+
+  public ConcurrentHashMap<String, AuditConfig> getAuditConfigs() {
+    return auditConfigs;
+  }
+
+  /**
+   *  create and enqueue a LoggingAuditEvent if TopicAuditConfig exists for this topic/logName,
+   *  enqueueEnabled is true and there is capacity available in the queue.
+   * @param loggingAuditHeaders
+   * @param messageValid
+   * @param messageAcknowledgedTimestamp
+   */
+  public void audit(String loggingAuditName, LoggingAuditHeaders loggingAuditHeaders,
+                    boolean messageValid, long messageAcknowledgedTimestamp) {
+    audit(loggingAuditName, loggingAuditHeaders, messageValid, messageAcknowledgedTimestamp, "",
+        "");
+  }
+
+  /**
+   *  create and enqueue a LoggingAuditEvent if TopicAuditConfig exists for this topic/logName,
+   *  enqueueEnabled is true and there is capacity available in the queue.
+   * @param loggingAuditName
+   * @param loggingAuditHeaders
+   * @param messageValid
+   * @param messageAcknowledgedTimestamp
+   * @param kafkaCluster
+   * @param topic
+   */
+  public void audit(String loggingAuditName, LoggingAuditHeaders loggingAuditHeaders,
+                    boolean messageValid,
+                    long messageAcknowledgedTimestamp, String kafkaCluster, String topic) {
+    if (!this.auditConfigs.containsKey(loggingAuditName)) {
+      OpenTsdbMetricConverter
+          .incr(LoggingAuditClientMetrics.AUDIT_EVENT_WITHOUT_TOPIC_CONFIGURED_ERROR,
+              "logName=" + loggingAuditHeaders.getLogName(), "host=" + host,
+              "stage=" + stage.toString());
+      return;
+    }
+    if (enqueueEnabled.get()) {
+      LoggingAuditEvent loggingAuditEvent = loggingAuditEventGenerator.generateAuditEvent(
+          loggingAuditName, loggingAuditHeaders, messageValid, messageAcknowledgedTimestamp,
+          kafkaCluster, topic);
+      boolean successful = false;
+      try {
+        // compared to put() which is blocking until available space in the queue, offer()
+        // returns false for enqueuing object if there is still no space after waiting for the
+        // specified the time period.
+        successful =
+            queue.offer(loggingAuditEvent, enqueueWaitInMilliseconds, TimeUnit.MILLISECONDS);
+      } catch (InterruptedException e) {
+        OpenTsdbMetricConverter
+            .incr(LoggingAuditClientMetrics.AUDIT_CLIENT_ENQUEUE_EXCEPTION, "host=" + host,
+                "stage=" + stage.toString());
+      }
+      if (successful) {
+        int qSize = queue.size();
+        double qUsagePercent = qSize * 1.0 / (qSize + queue.remainingCapacity());
+        OpenTsdbMetricConverter
+            .gauge(LoggingAuditClientMetrics.AUDIT_CLIENT_QUEUE_SIZE, qSize, "host=" + host,
+                "stage=" + stage.toString());
+        OpenTsdbMetricConverter
+            .gauge(LoggingAuditClientMetrics.AUDIT_CLIENT_QUEUE_USAGE_PERCENT, qUsagePercent,
```

Comment:
if we have metrics on `AUDIT_CLIENT_QUEUE_USAGE_PERCENT`, do we still need `AUDIT_CLIENT_QUEUE_SIZE`?

Reply:
Queue capacity is a configuration property for LoggingAuditClient. For different use cases with different QPS,  queue capacity can be very different.   AUDIT_CLIENT_QUEUE_USAGE_PERCENT and  AUDIT_CLIENT_QUEUE_SIZE together can show the status of the queue and this will help to tune or set the queue_capacity property.   Therefore, I think both metrics are useful.